Add Rxjs and an interaction layer

rogerhutchings opened the pull request:
Package: lib-classifier

Adds Rxjs, and adds an interaction layer for recording mousemove, mouseup and mousedown, and publishes those events to a stream. The stream currently logs to console, but in a future PR that'll be used for handling annotation creation.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?
- [ ] Is the changelog updated?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?

srallen commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayerContainer.js:
```diff
@@ -0,0 +1,70 @@
+
+  componentDidMount () {
+    const stream = from(toStream(() => this.props.drawing.mouseEventStream))
+    stream.subscribe(z => console.log(z))
```
Maybe leave an inline to do comment if there's going to be a defined callback here later?

rogerhutchings replied:
I've added one, but it's not showing up in this comment diff - you can see it at https://github.com/zooniverse/front-end-monorepo/pull/170/files#diff-b5aee330efdbec039e4275348c0a1bc0R27

srallen commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayerContainer.js:
```diff
@@ -0,0 +1,70 @@
+
+
+  componentDidMount () {
+    const stream = from(toStream(() => this.props.drawing.mouseEventStream))
```
I'm not very well versed in rx.js's API. Just wondering why `from` is used here instead of `fromEvent` (https://rxjs-dev.firebaseapp.com/api/index/function/fromEvent) since what we're tracking are DOM events.

rogerhutchings replied:
I think `fromEvent` is a shorthand for binding event listeners, so we could use that instead of the React event handlers lower down. We'd then have to do cleanup ourselves when the component gets unmounted. Also, I know React also has some optimisations under the hood, like reusing event objects; I'm not sure if Rxjs does anything similar.

rogerhutchings replied:
In retrospect, I'm not sure this is the most idiomatic way of doing it though...

srallen commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/SubjectViewer.js:
```diff
@@ -30,9 +37,14 @@ class SubjectViewer extends React.Component {
+    const { className, subject } = this.props
     const Viewer = getViewer(subject.viewer)
-    return <Viewer subject={subject} {...this.props} />
+    return (
```
Having the SVG layer on always on top regardless of the viewer type is what breaks native HTML 5 video player controls, which is especially annoying on mobile devices. We've run into this problem on PFE with the video projects. I don't think we've had a project ask for SVG drawing annotations on video yet and something like that might be hard to implement for annotations. 

I think the text subject types that use the highlighter tool also don't need an SVG layer.

rogerhutchings replied:
Fair enough - I'll defer to the individual subject viewer on how it renders then.

rogerhutchings commented:
I've addressed the inline comments, simplified the stream creation - now, it's simply available as an object you can `subscribe` to in the store - and added some specs.

eatyourgreens commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewer.js:
```diff
@@ -1,85 +1,25 @@
+  return (
+    <SVG>
+      <image xlinkHref={url} />
+      <InteractionLayer />
```
Should there be some conditional logic around the interaction layer? In PFE, the SVG doesn't respond to mouse/touch events unless you're drawing so is there some coupling with the current task here?

rogerhutchings replied:
I was going to `return null` in the InteractionLayer if drawing isn't enabled.

rogerhutchings commented on packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/SingleImageViewer/SingleImageViewerContainer.js:
```diff
@@ -0,0 +1,85 @@
+      this.setState({
+        height: img.height,
+        width: img.width,
+        loading: asyncStates.loading
```
Good catch

rogerhutchings commented:
I was going to start a separate branch for the whole drawing API, but that probably isn't necessary. I'll merge to master.
